pkg/lib/verifier/rule_builder_util.go

```diff
@@ -17,112 +18,174 @@ import (
 	"github.com/pkg/errors"
 )
 
-func getPcrMatchesConstantRules(pcrs []types.PcrIndex, flavor *hvs.Flavor, marker common.FlavorPart) ([]rules.Rule, error) {
+//getPcrMatchesConstantRules method will create PcrMatchesConstantRule and return the rule
+//return nil if error occurs
+func getPcrMatchesConstantRules(pcrs []types.PcrIndex, flavor *hvs.Flavor, pcrLogData *types.PCRS, marker common.FlavorPart) ([]rules.Rule, error) {
 
 	var results []rules.Rule
+	var rule rules.Rule
+	var err error
 
-	// iterate over the banks, collecting the values for each supplied index
-	// and create PcrMatchesConstant rules.
-	for bank, pcrMap := range flavor.Pcrs {
-		for _, index := range pcrs {
-			if expectedPcrEx, ok := pcrMap[index.String()]; ok {
-				expectedPcr, _ := rules.FlavorPcr2ManifestPcr(&expectedPcrEx, types.SHAAlgorithm(bank), index)
+	if pcrLogData != nil {
```

Comment:
During Esxi flow, we are passing the pcrLogData as nil

Reply:
@kwtj43 I think we must not have two separate flows, let us know your view on this.

Reply:
All of the functions in rule_builder_util.go are "helper" functions that simplify the creation of "rules" (in this case the PCR Matches Constant rule).  Each instance of that rule compares a single PCR value from the flavor (say at index 2) against the value from the host manifest.  Previously, things were hard coded -- for example if SUEFI was enabled, we would enforce additional PCRs (ex. 1 - 7).  There is no "log data" that is evaluated in PCR Matches Constant rules (just final values of the PCRs).  With this enhancement we expect that the flavor will contain the PCR matches logic (this function may not be needed anymore).  For example, if the flavor contains "pcr_matches: true" for index 2, we should create the rule (with the flavor's value).  It will later be evaluated during "Apply()".  I hope that clarifies things -- if not, let's find time to meet to discuss the logic.

Reply:
@ddhawal  @kwtj43 
In the existing implementation, Pcr indices are hard coded and passed to the rule builder functions. Whereas in the new implementation, we are not passing the index to fetch the Pcr value. Rather than we pass the whole pcr log data which is of the below struct type.
type PCRS struct {
	PCR              PCR                `json:"pcr"`         //required
	Measurement      string             `json:"measurement"` //required
	PCRMatches       bool               `json:"pcr_matches,omitempty"`
	EventlogEqual    *EventLogEqual     `json:"eventlog_equals,omitempty"`
	EventlogIncludes []EventLogCriteria `json:"eventlog_includes,omitempty"`
}

Note(rule_factory.go) : This pcr log data was taken from the flavor part by parsing the rule field name. (in this case it is "PCRMatches"). 

In order to maintain the backward compatibility, we need both the cases.